### DISCUSSION TITLE
Chart aspect ratio / calculating angles

### DISCUSSION THREAD

**Discussion by M22arius494** [DaveSkender/Stock.Indicators#726] [Q&A] [answered]
Hello,

I'm trying to calculate the angles of the EMA50 / EMA20 etc. Just using the deltax and deltay as shown below. That kind of works, but i'm running into troubles because of the (aspect)ratio of crypto charts. A lot is depending on a proper alignment of the x-axis (date) and y-axis (price). In tradingview it is possible to propertly "lock" (not sure if that is the right word) the aspect ratio for use of the "Gan Fan").

Any idea's how to calculate the aspect ratio or how such can be done (or how tradingview calculates it?)?

Thanks,
Marius

//(0,0) -> (10,10)
double angle1 = Math.Atan2(10 - 0, 10 - 0);
double angle1Degrees = ConvertRadiansToDegrees(angle1);

//(0,0) -> (10,5)
double angle2 = Math.Atan2(5 - 0, 10 - 0);
double angle2Degrees = ConvertRadiansToDegrees(angle2);

double resultDiff = angle1 - angle2;
double resultDegrees = ConvertRadiansToDegrees(resultDiff);

**Comment by DaveSkender** [comment]
You can explore [Slope](https://daveskender.github.io/Stock.Indicators/indicators/Slope/#content).  This uses a Linear Regression best-fit line approach (could be over 2 periods).  If you recall the formula, `y=m*x+b`, where `m` is the slope.  See also #630 -- we'll be looking at making this easier to apply to any indicator.  For now, you'd have to [convert to quote](https://daveskender.github.io/Stock.Indicators/utilities/#convert-to-quotes) first.

**Comment by DaveSkender** [comment, marked as answer]
I’m looking at [Gan Fan](https://www.investopedia.com/terms/g/gann-fans.asp) and am confused by it.  It seems to be irrational since it bases it’s thinking on one unit up and one unit over is 45* angle, but fails to define one unit up in any consistent way.  $1 up is very different between a $5 price and a $500 price and 1% isn’t any better.  For this to make sense, both x and y axis would need to be using the same units.

**Reply by DaveSkender** [reply]
To get this to work, I think you’d have to make one unit up = one standard deviation in price, or ATR.  Guessing.

**Reply by DaveSkender** [reply]
And yes, you’re right about chart aspect ratio, plotting 45* would only look right with a very specific x:y ratio of 1:1, based on whatever you choose as the “one up” unit.

First glance, I’d say Gan Fan is not a calculable indicator at all; rather it’s something you drop onto a chart, adjust to desired “one up” value (a user provided input), and just visually adjust until it looks nice.

**Reply by M22arius494** [reply]
> And yes, you’re right about chart aspect ratio, plotting 45* would only look right with a very specific x:y ratio of 1:1, based on whatever you choose as the “one up” unit.
> 
> First glance, I’d say Gan Fan is not a calculable indicator at all; rather it’s something you drop onto a chart, adjust to desired “one up” value (a user provided input), and just visually adjust until it looks nice.

That is right, it is "just" a tool to "predict" future price development (and it amazed me also). I don't really use it much but I know it needs the proper scaling to predict.

45 degrees is only proper if the distance between date (x-axis) and price (y-axis) have the same distance . In tradingview you can see that via settings -> scales -> lock and you can see an ratio-value there. I was after the logic on how to calculate that ratio. 

Without a proper ratio a slope doesn't say anything. In chart A the slope can be totally different than in chart B leading to "confusing" for automated bot decisions.

**Reply by DaveSkender** [reply]
I think to get a true 45* visually, you have to take into account the high high and low low across the visible chart to get the Y units then proportionally adjust for "one unit", then limit the number of bars accordingly shown on the X axis.
